### DISCUSSION TITLE
Contribution towards Emscripten + Wasi

### DISCUSSION THREAD

**Discussion by jeff-hykin** [wasmerio/wasmer#2623] [Q&A]
### Summary

Is it practical for a outsider/non-major-contributor to implement the emscripten + wasi environment ([relevent code snippet here](https://github.com/wasmerio/wasmer/blob/7d491b4bb3c5121208df1f8394e968b08116d447/lib/cli/src/commands/run.rs#L129))?

### Additional details

Current plan is to combine this [wasi env](https://github.com/wasmerio/wasmer/blob/46024fbc2f9c5cc1810a3cc312d2c287f9f811f3/lib/wasi/src/lib.rs#L181) with the equivlent [emscripten env](https://github.com/wasmerio/wasmer/blob/861fe48096d02264f804d071fa1102e9f09b0dbd/lib/emscripten/src/lib.rs#L558) and just rename them to match the `___wasi_` naming 

I only need it for a specific usecase, not a stable feature. Someone compiled opencv to wasm and I just want to get that working at an alpha version in Rust. I know some about implementing wasm imports in JS, which wasnt too difficult, but I don't know much about the internal structure of the wasmer crate or the wasmer runtime.

**Comment by Hywan** [comment]
Hello,

Thanks for the question.
Emscripten and WASI are really different ABI, and that's why Wasmer doesn't support both _at the same time_. Wasmer support Emscripten, or WASI, but separately.

Does your opencv project have been compiled to Wasm with Emscripten only?

**Comment by jeff-hykin** [comment]
Its not mine, I just pulled the wasm file out from [here](https://github.com/echamudi/opencv-wasm)

It is definitely Emscripten as the main ABI, but it has a whole page of `____wasi__*function name*` imports like file descriptor close, alloc, etc

**Comment by Hywan** [comment]
Well OK, I'm sorry but it's not on our agenda to support both ABI for the moment. As far as I know, it's rather unusual to get two of them at the same time (it's likely that I'm wrong here!).

Contributions are welcomed as it was your first question! I suppose it's doable, as we use the same `ImportObject` API. We can generate a first `ImportObject` with the WASI API, then generate second `ImportObject` with the Emscripten API, to finally merge both of them into a single `ImportObject`. I don't see any particular blocker, except maybe the way we manage the different environements, but I believe it's compartmentalised per imports, so it _may_ not be an issue. @MarkMcCaskey can confirm that.

**Comment by jeff-hykin** [comment]
Okay thanks! I'll look for Mark's reply and if there's no obvious reason why it would fail I'll slowly move along on this feature.

Maybe there's a more simple way to do it like you're describing (I'm not sure how to combine import objects). I've just duplicated the Emscripten folder, and been merging the WASI folder files/code into it. It's been pretty straightforward, the only thing they have conflicting definitions of is WasmPtr, and just renaming to "WasiWasmPtr" will hopefully resolve that. There's some semi-complicated module import issues I'm working on, but that's more of a Rust issue/complication.

I agree I don't know why the API's are mixed together, but that link in the first post to the dedicated error message for this situation I guess suggests it's somewhat common.

**Comment by jcaesar** [comment]
> As far as I know, it's rather unusual to get two of them at the same time

Not really: https://github.com/wasmerio/wasmer/issues/1977#issuecomment-754747526 (Just linking this to organize related issues. I suspect you already know.)
